### PR TITLE
raftstore: make it possible to apply recursively

### PR DESCRIPTION
**Pull request by BusyJay**
To apply recursively, use `stash` first, and then call `handle_raft_committed_entries`, finally `restore_stash`.

/cc #2502

### CODE REVIEW

**Comment by huachaohuang** [comment]
Can you give an example of why and when we need to apply recursively?

**Comment by BusyJay** [comment]
For example, B is supposed to be merged into A, then A is supposed to catch up all logs on behalf of B if it's still behind leader's progress. So A needs to stash current changes, apply all B's entries, and then restore previous changes, finally execute merge.

File: src/raftstore/store/worker/apply.rs [hunk 1]
```diff
@@ -1543,33 +1611,35 @@ impl Runner {
...
             .apply_res_capacity(applys.len())
             .use_delete_range(self.use_delete_range)
             .enable_sync_log(self.sync_log);
         for apply in applys {
             if apply.entries.is_empty() {
                 continue;
             }
-            let mut e = match self.delegates.entry(apply.region_id) {
-                MapEntry::Vacant(_) => {
+            let mut delegate = match self.delegates.get_mut(&apply.region_id) {
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: No, `handle_raft_committed_entries` needs to borrow `self.delegates` exclusively.

**Comment by BusyJay** [comment]
PTAL

File: src/raftstore/store/worker/apply.rs [hunk 2]
```diff
@@ -2034,10 +2114,11 @@ mod tests {
         let obs = ApplyObserver::default();
         host.registry
             .register_query_observer(1, Box::new(obs.clone()));
-        let mut apply_ctx = ApplyContext::new(&host).use_delete_range(true);
+        let core = ApplyContextCore::new(&host).use_delete_range(true);
+        let mut apply_ctx = ApplyContext::new(core, &mut delegates);
         delegate.handle_raft_committed_entries(&mut apply_ctx, vec![put_entry]);
-        apply_ctx.write_to_db(&db);
-        assert!(apply_ctx.apply_res.last().unwrap().exec_res.is_empty());
+        apply_ctx.core.write_to_db(&db);
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: `ApplyContext` will deref to `ApplyContextCore`, so don't need to change here?

File: src/raftstore/store/worker/apply.rs [hunk 3]
```diff
@@ -2058,9 +2139,10 @@ mod tests {
             .put_cf(CF_LOCK, b"k1", b"v1")
             .epoch(1, 3)
             .build();
-        let mut apply_ctx = ApplyContext::new(&host).use_delete_range(true);
+        let core = ApplyContextCore::new(&host).use_delete_range(true);
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: Can we reuse `core` or `apply_ctx` here?

File: src/raftstore/store/worker/apply.rs [hunk 4]
```diff
@@ -1543,33 +1611,35 @@ impl Runner {
...
+                core = ctx.core;
             }
-            if e.get().pending_remove {
-                e.remove();
+
+            if delegate.pending_remove {
+                delegate.destroy();
+                self.delegates.remove(&apply.region_id);
+            } else {
+                *self.delegates.get_mut(&apply.region_id).unwrap() = Some(delegate);
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: I'm not sure if insert will introduce allocation, but `*get_mut().unwrap() = ` certainly won't.

File: src/raftstore/store/worker/apply.rs [hunk 5]
```diff
@@ -310,6 +323,29 @@ impl<'a> ApplyContext<'a> {
...
+            region: self.cbs.last().map(|cbs| cbs.region.clone()),
+            exec_ctx: self.exec_ctx.take(),
+            last_applied_index: self.last_applied_index,
+        }
+    }
+
+    /// Restore the dirty state, so context can resume applying from
+    /// last stash point.
+    pub fn restore_stash(&mut self, stash: Stash) {
+        if let Some(region) = stash.region {
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: When will this be `None`?

File: src/raftstore/store/worker/apply.rs [hunk 6]
```diff
@@ -310,6 +323,29 @@ impl<'a> ApplyContext<'a> {
...
+            self.cbs.push(ApplyCallback::new(region));
```
Comment (huachaohuang) [inline src/raftstore/store/worker/apply.rs]: If we `stash` and `restore` the same region in the same `ApplyContextCore`, the callback of the same region will be called twice, is that OK?

File: src/raftstore/store/worker/apply.rs (hunk 5 above)
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: When `stash` is callced before `prepare_for`.

File: src/raftstore/store/worker/apply.rs (hunk 6 above)
Comment (BusyJay) [reply src/raftstore/store/worker/apply.rs]: No, here a new `ApplyCallback` is created. So the same callback won't be called twice.

**Pull request merged**